### PR TITLE
Add option to resize Cover Block

### LINKED ISSUE #17113: Allow "Cover" block to be resizable
It'd be interesting to extend the **Cover** block with `ResizableBox`, in the same implementation as the **Spacer** block (resize handle at the bottom center), to allow changing the default height of the block.

### PR DESCRIPTION
**Pull request by senadir**
## Description
closes #17113
This PR adds resizing to Cover block & an option to manually select the value.

after #17113 is merged, I need to rebase this PR around it.

## How has this been tested?
* create a new cover block
* resize
* save

## Screenshots 

cover-resize

### CODE REVIEW

**Comment by jasmussen** [comment]
MY GOLLY!

I don't have time to review this today, will look tomorrow. But I think @kjellr wants to look at this.

Even from the GIF it also seems clear that we'll want to adjust how the sibling inserter works here. In a related case, a good friend of mine suggested that maybe we adjust the sibling inserter (the plus that appears on hover) to not appear until 2 seconds has passed, something in that vein. Another option is to disable it when the block above has resize handles. Or is selected. There are a few ways to go here, but it feels like a key issue to solve before we roll out such a vertical handle.

**Comment by senadir** [comment]
@mapk have a PR (https://github.com/WordPress/gutenberg/pull/17136) related to this problem so I didn't want to address it here but yeah, the bottom resizer is being shadowed by the inserter.
I'm not sure about the delay, delays works great in visual animations but not functional animations, I think the spacing should be fine?

**Comment by mtias** [comment]
There's some more ideas here for the sibling inserter. I don't think spacing alone is enough: https://github.com/WordPress/gutenberg/issues/16646#issuecomment-523580520

**Comment by ZebulanStanphill** [comment]
Wouldn't it make more sense for the drag handle and input to control the `min-height` of the block, to prevent the inner content from overflowing out of the container when squeezed down to small widths?

File: packages/block-library/src/cover/edit.js [hunk 1]
```diff
@@ -268,39 +296,67 @@ class CoverEdit extends Component {
...
+						right: false,
+						bottom: true,
+						left: false,
+						topRight: false,
+						bottomRight: false,
+						bottomLeft: false,
+						topLeft: false,
+					} }
+					onResizeStop={ ( event, direction, elt, delta ) => {
+						const coverHeight = parseInt( height, 10 ) + parseInt( delta.height, 10 );
```
[the remark this answers is not in the document]
Comment (senadir) [reply packages/block-library/src/cover/edit.js]: for some reason this caused some issues for me, giving me 42015 rather then 435
I will try again & report back

File: packages/block-library/src/cover/save.js [hunk 2]
```diff
@@ -41,7 +42,9 @@ export default function save( { attributes } ) {
 	if ( focalPoint && ! hasParallax ) {
 		style.backgroundPosition = `${ focalPoint.x * 100 }% ${ focalPoint.y * 100 }%`;
 	}
-
+	if ( backgroundType === IMAGE_BACKGROUND_TYPE && height ) {
```
[the remark this answers is not in the document]
Comment (senadir) [reply packages/block-library/src/cover/save.js]: yes, my initial intention was to apply the resizing to image only but I changed my mind later & forget to refactor this.

File: packages/block-library/src/cover/block.json [hunk 3]
```diff
@@ -28,6 +28,9 @@
 		},
 		"focalPoint": {
 			"type": "object"
+		},
+		"height": {
+			"type": "number"
```
Comment (mtias) [inline packages/block-library/src/cover/block.json]: Can we add the default height here? `"default": 430`

Comment (youknowriad) [reply packages/block-library/src/cover/block.json]: Adding the default here will invalidate previous blocks. We'll need a deprecation if we do so.

**Comment by senadir** [comment]
> Wouldn't this be better if it changed the min-height rather than height? Imagine mobile viewports where the content spans more lines because of less horizontal space.
> 
> Also, does the `COVER_MIN_HEIGHT` value need to be hardcoded? I'm already using this block on some sites with a much lower min-height value. Usually because they're presented in columns. I guess that means if a user so much as touches the resizer, it will get a height inline style value that's impossible to remove.

those are some very good points @oxyc that I haven't consider.

A second implementation that I didn't upload yet, changes height to min-height, as for the hardcoded min height, I guess I should differ to something more absolute, 20 or 50px sounds reasonable to not get a invisible block and still over wide arrangements.

after giving the code a second look, and with @jorgefilipecosta review, I guess I was just too damn tired yesterday 😅

File: packages/block-library/src/cover/edit.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (senadir) [reply packages/block-library/src/cover/edit.js]: never mind it seems I was setting the initial value to `''` causing it to be a string.

File: packages/block-library/src/cover/edit.js [hunk 4]
```diff
@@ -268,39 +296,67 @@ class CoverEdit extends Component {
...
+						setAttributes( {
+							height: coverHeight,
```
[the remark this answers is not in the document]
Comment (senadir) [reply packages/block-library/src/cover/edit.js]: I moved some of the logic away from the resize and kept everything in the input

**Comment by senadir** [comment]
I've changed the behavior, 430 is not the min height now but the default one. in the future we might abstract it even more to be a theme option, if there is a better option for now I'm all ears to implement it. 

the new min value is arbitrarily set to 50px, no strong feeling toward this value, just seemed small enough to not restrict and big enough so the cover is not invisible.

Also now we change the min height & not the height, should fix a lot of compatibility problems.

the input logic has been heavily inspired (read: literally copied & changed variable names) from the spacer block https://github.com/WordPress/gutenberg/pull/14785, thank you @Jackie6  & @talldan for the awesome work solving all of those problems for me before hand.

I think this is ready for a final review

**Comment by jasmussen** [comment]
Nice work.

Here's what I'm seeing:

resize

When you increase the vertical height, it looks good. When you decrease the vertical height, it appears as if the content inside only "snaps into place" when you release the mouse. Can we do so increasing and decreasing size both work smoothly?

**Comment by senadir** [comment]
@jasmussen you know, this bug bothered me for the past 4 days and I didn't quite knew how to fix, but then you mentioned it and suddenly I knew the source of the problem & how to fix. Please comment more on my problems 😂
fixed in 8fd64f775
resizing-block-fixed

**Comment by jasmussen** [comment]
Solid work!

**Comment by kjellr** [comment]
This looks great. 🎉

Just a quick coordination note: we'll want to make sure that this plays nicely with the `100vh` "Fullscreen" option under development in https://github.com/WordPress/gutenberg/pull/16738 (cc @retrofox). My guess is that when the fullscreen option is selected, these controls would disappear or be disabled automatically.

**Review by jorgefilipecosta — APPROVED** [review]
Excellent work @senadir 👍 The code changes look good!

Here are some possible improvements I noticed:
- We are setting a min-height, and it works great on the frontend on the editor thought it looks as if we are setting a height. If the content is bigger than the height, we set the height does not expand. We may need to configure the style generation of resizeable (something similar to what happened with media & text).
- If I resize a cover block, a min-height will always be set from now on. There is nothing I can do to avoid setting a min-height after applying some resizing operation. We probably need some mechanism to allow a reset.

**Pull request merged**